A51-xds-custom-backend-metrics.md

```diff
@@ -0,0 +1,382 @@
+To do that, we should not compare maps for de-duplication, instead, we can bump up a 
+generation_id of the metrics data every time users make a change. 
+Each client connection only needs to compare its previous generation_id to do deduplication. 
+The optimization strategy can be implemented until there are concrete use cases in the future.
+
+#### The Client API 
+
+Out-of-band data will be sent only if the client starts a streaming call for the `OpenRCAService/StreamCoreMetrics` method.
+If the backend is designed to provide this information but the client never asks for it, it will not be used.  
+Conversely, if the backend does not provide the information but the client does start a call asking for it,
+the backend will terminate the call with status `UNIMPLEMENTED`, in which case the client will not retry the call.
+
+Any LB policy that is interested in this out-of-band data may subscribe to it.  
+The client will maintain an open `OpenRCAService/StreamCoreMetrics` call 
+whenever there is at least one LB policy subscribing to the data.
+
+When a policy subscribes, it will indicate the interval at which it would like to get updates. 
+If multiple policies subscribe with different intervals, we will use the minimum of those intervals.  
+Note that the ORCA protocol involves the client telling the server the interval at which 
+it wants reports when the call starts, so if the client needs to change its requested interval, 
+it will need to cancel and restart the call.
+A client may change its requested interval at any time; this is expected to be a fairly rare operation 
+(usually triggered by a config change for one of the subscribing LB policies),
+It is important that a load balancing policy is able to re-config the frequency
+and should not require heavy-weight operations like reconnecting to backends or recreating subchannels or LB policies.
+
+Each policy that has subscribed to the data will get a callback whenever a new message is sent to the client on the call.  
+As with per-request data, the protobuf deserialization will happen only once.  
+The data provided to the callback should be exposed to the LB policy using the same type as used for per-request data.
+
+The API for subscribing to out-of-band reports must be available to LB policies at any level of the routing hierarchy, 
+including both leaf LB policies (whose children are subchannels) and parent LB policies (whose children are other LB policies).  
+Note that there may be multiple LB policies in the routing hierarchy that are interested in subscribing.  
+For example, in the case where we are using the xDS policy for cluster routing and the WRR policy for intra-cluster routing,
+both policies may be interested in subscribing.  
+The WRR policy will be directly managing the subchannels and will therefore be able to subscribe directly against the subchannel API, 
+whereas the xDS policy will need to subscribe against the LB policy API of the WRR policy, 
```

Comment:
I think we decided that this isn't the way we're actually going to do it.  Instead of having a way for parent policies to subscribe via the child policy (which would require adding new methods to the LB policy API itself), the parent policy will just subscribe to the subchannels as they are created before passing them back down to the child policy.

@ejona86 @dfawley Please confirm that this matches your recollection of where we left this.

Reply:
Thank you. Did we decide to do a wrapper `ForwardingLoadBalancerHelper `?  c.f. [java implementation](https://github.com/grpc/grpc-java/blob/881747a63df79a7baae252de86447c0997c76f69/xds/src/main/java/io/grpc/xds/OrcaOobUtil.java#L97-L147).

Reply:
Yeah, the text at the time of this comment doesn't seem likely in any language. I see the text changed for this already, and is now mirroring the Java approach. I think though that will probably need to go into the Java section and we leave this part here a bit more generic (just move the "We will provide a utility function... accepts subscriptions from any routing hierarchy." part to Java and leave the rest as-is?)

Last I heard, C was special-casing this in their subchannel API, so a parent can use the API on subchannel creation before returning a subchannel to a child. I thought Go was considering something similar, although that was partly because of the restrictions at the time; maybe Go has more options now.

Reply:
Yeah, the new text here is true of Java but not true in other languages -- C-core will not do this by wrapping the helper at all.

I suggest just removing the paragraph you added here (line 218 in the most recent commit).  I think there's probably enough context in the paragraphs above to make the context clear.